doc/SAI-Proposal-RIF-Metadata.md

```diff
@@ -0,0 +1,362 @@
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_ROUTER_INTERFACE_ATTR_META_DATA_INGRESS,
+
+    /**
+     * @brief User based Meta Data Egress
+     *
+     * Value Range #SAI_ACL_CAPABILITY_EXTENDED_TYPE_ROUTER_INTERFACE_META_DATA for Egress stage.
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_ROUTER_INTERFACE_ATTR_META_DATA_EGRESS,
+    ...
+} sai_router_interface_attr_t;
+```
+
+### ACL Table and Entry Match Field Additions for RIF Metadata
+
+`saiacl.h`
+```c
+typedef enum _sai_acl_table_attr_t
+{
+    ...
+    /**
+     * @brief Router Interface User metadata
+     *
+     * @type bool
+     * @flags CREATE_ONLY
+     * @default false
+     */
+    SAI_ACL_TABLE_ATTR_FIELD_ROUTER_INTERFACE_USER_META = SAI_ACL_TABLE_ATTR_FIELD_START + 0x162,
+    ...
+} sai_acl_table_attr_t;
+
+typedef enum _sai_acl_entry_attr_t
+{
+    ...
+    /**
+     * @brief Router Interface User metadata
+     *
+     * Value must be in the range defined in
+     * SAI_ACL_CAPABILITY_EXTENDED_TYPE_ROUTER_INTERFACE_META_DATA
+     * SAI_ROUTER_INTERFACE_ATTR_META_DATA_INGRESS in SAI_ACL_STAGE_POST_INGRESS stage
+     * or SAI_ACL_CAPABILITY_EXTENDED_TYPE_ROUTER_INTERFACE_META_DATA for
+     * SAI_ROUTER_INTERFACE_ATTR_META_DATA_EGRESS in SAI_ACL_STAGE_EGRESS stage
+     *
+     * @type sai_acl_field_data_t sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default disabled
+     */
+    SAI_ACL_ENTRY_ATTR_FIELD_ROUTER_INTERFACE_USER_META = SAI_ACL_ENTRY_ATTR_FIELD_START + 0x162,
+    ...
+} sai_acl_entry_attr_t;
+```
+
+### Switch Attributes for Extended ACL Capability Query per ACL stage
+
+`saiswitch.h`
+```c
+typedef enum _sai_switch_attr_t
+{
+    ...
+    /**
+     * @brief Pre Ingress ACL stage.
+     *
+     * @type sai_acl_capability_extended_list_t
+     * @flags READ_ONLY
+     */
+    SAI_SWITCH_ATTR_PRE_INGRESS_ACL_CAPABILITY_EXTENDED,
+    /**
+     * @brief Ingress ACL stage.
+     *
+     * @type sai_acl_capability_extended_list_t
+     * @flags READ_ONLY
+     */
+    SAI_SWITCH_ATTR_INGRESS_ACL_CAPABILITY_EXTENDED,
+    /**
+     * @brief Egress ACL stage.
+     *
+     * @type sai_acl_capability_extended_list_t
+     * @flags READ_ONLY
+     */
+    SAI_SWITCH_ATTR_EGRESS_ACL_CAPABILITY_EXTENDED,
+    ...
+} sai_switch_attr_t;
+```
+
+### ACL Capability Extended Query Types and Structures
+
+`saitypes.h`
+
+```c
+/**
+ * @brief Field type for ACL capability extended
+ */
+typedef enum _sai_acl_capability_extended_type_t
+{
+    /** Router Interface user-based meta data range */
+    SAI_ACL_CAPABILITY_EXTENDED_TYPE_ROUTER_INTERFACE_META_DATA,
+
+} sai_acl_capability_extended_type_t;
+
+/**
+ * @extraparam sai_acl_capability_extended_type_t type
+ */
+typedef union _sai_acl_capability_extended_value_t
+{
+    /** @validonly type == SAI_ACL_CAPABILITY_EXTENDED_TYPE_ROUTER_INTERFACE_META_DATA */
+    sai_u32_range_t router_interface_meta_data_range;
```

Comment:
what happens if we want to add additional attributes in the future and need a bigger data type. This becomes again non binary compatible